cpp/src/arrow/compute/exec/project_benchmark.cc

```diff
@@ -0,0 +1,141 @@
+
+static void ProjectionOverhead(benchmark::State& state, Expression expr) {
+  const auto batch_size = static_cast<int32_t>(state.range(0));
+  const auto num_batches = total_batch_size / batch_size;
+
+  auto data = MakeRandomBatches(schema({field("i64", int64()), field("bool", boolean())}),
+                                num_batches, batch_size);
+  ExecContext ctx(default_memory_pool(), arrow::internal::GetCpuThreadPool());
+  for (auto _ : state) {
+    state.PauseTiming();
+    ASSERT_OK_AND_ASSIGN(auto plan, ExecPlan::Make(&ctx));
+    AsyncGenerator<util::optional<ExecBatch>> sink_gen;
+    ASSERT_OK(Declaration::Sequence(
+                  {
+                      {"source",
+                       SourceNodeOptions{data.schema,
+                                         data.gen(/*parallel=*/true, /*slow=*/false)},
+                       "custom_source_label"},
+                      {"project", ProjectNodeOptions{{
+                                      expr,
+                                  }}},
+                      {"sink", SinkNodeOptions{&sink_gen}, "custom_sink_label"},
+                  })
+                  .AddToPlan(plan.get()));
+    state.ResumeTiming();
+    ASSERT_FINISHES_OK(StartAndCollect(plan.get(), sink_gen));
+  }
+  state.counters["rows_per_second"] = benchmark::Counter(
+      static_cast<double>(state.iterations() * num_batches * batch_size),
+      benchmark::Counter::kIsRate);
+
+  state.counters["batches_per_second"] = benchmark::Counter(
+      static_cast<double>(state.iterations() * num_batches), benchmark::Counter::kIsRate);
+}
+
+static void ProjectionOverheadIsolated(benchmark::State& state, Expression expr) {
+  const auto batch_size = static_cast<int32_t>(state.range(0));
+  const auto num_batches = total_batch_size / batch_size;
+
+  auto data = MakeRandomBatches(schema({field("i64", int64()), field("bool", boolean())}),
+                                num_batches, batch_size);
+  ExecContext ctx(default_memory_pool(), arrow::internal::GetCpuThreadPool());
+  for (auto _ : state) {
+    state.PauseTiming();
+    AsyncGenerator<util::optional<ExecBatch>> sink_gen;
+
+    ASSERT_OK_AND_ASSIGN(auto plan, ExecPlan::Make(&ctx));
+
+    ASSERT_OK_AND_ASSIGN(
+        auto sn, MakeExecNode("source", plan.get(), {},
+                              SourceNodeOptions{data.schema, data.gen(/*parallel=*/true,
+                                                                      /*slow=*/false)}));
+    ASSERT_OK_AND_ASSIGN(auto pn, MakeExecNode("project", plan.get(), {sn},
+                                               ProjectNodeOptions{{
+                                                   expr,
+                                               }}));
+    MakeExecNode("sink", plan.get(), {pn}, SinkNodeOptions{&sink_gen});
+    pn->InputFinished(sn, num_batches);
+    state.ResumeTiming();
+    for (auto b : data.batches) {
+      pn->InputReceived(sn, b);
+    }
+    pn->finished();
```

Comment:
Is this the correct final condition, e.g Is this a blocking call until finished is true?

Reply:
You can probably get away with `ASSERT_TRUE(pn->finished().is_finished())`.  In the future (< 3 months maybe?) I'm guessing the entire concept of a single node finishing will go away.